File: bungee-impl-hidden-details/src/main/java/us/minevict/mvutil/common/utils/SetupPlatformless.java
package us.minevict.mvutil.common.utils;

import co.aikar.commands.BungeeCommandIssuer;
import co.aikar.commands.CommandIssuer;
import java.util.concurrent.CompletableFuture;
import java.util.function.BiConsumer;
import java.util.function.Function;
import java.util.function.Supplier;
import net.md_5.bungee.api.ProxyServer;
import net.md_5.bungee.api.chat.BaseComponent;
import net.md_5.bungee.api.plugin.Plugin;

/**
 * @since 3.6.0
 */
public class SetupPlatformless {
  private SetupPlatformless() {
  }

  /**
   * Set up the {@link Platformless} functionality.
   * <p>
   * This <b>must</b> be called (at least) once.
   * </p>
   *
   * @param plugin The plugin executing this.
   */
  public static void setup(Plugin plugin) {
    Platformless.messageSender = PlatformlessMessageSenderBungee.INSTANCE;
    Platformless.asyncRunner = new PlatformlessAsyncRunnerBungee(plugin);
  }

  public static class PlatformlessMessageSenderBungee implements
      BiConsumer<CommandIssuer, BaseComponent[]> {
    public static final PlatformlessMessageSenderBungee INSTANCE = new PlatformlessMessageSenderBungee();

    private PlatformlessMessageSenderBungee() {
    }

    @Override
    public void accept(CommandIssuer commandIssuer, BaseComponent[] baseComponents) {
      if (baseComponents == null || !(commandIssuer instanceof BungeeCommandIssuer)) {
        return;
      }

      ((BungeeCommandIssuer) commandIssuer).getIssuer().sendMessage(baseComponents);
    }
  }

  public static class PlatformlessAsyncRunnerBungee implements
      Function<Supplier, CompletableFuture> {
    private final Plugin plugin;

    public PlatformlessAsyncRunnerBungee(Plugin plugin) {
      this.plugin = plugin;
    }

    @SuppressWarnings({"rawtypes", "unchecked"})
    @Override
    public CompletableFuture apply(Supplier supplier) {
      CompletableFuture future = new CompletableFuture();
      ProxyServer.getInstance().getScheduler().runAsync(plugin, () -> {
        try {
          future.complete(supplier.get());
        } catch (Throwable throwable) {
          future.completeExceptionally(throwable);
        }
      });
      return future;
    }
  }
}

File: bungee/src/main/kotlin/us/minevict/mvutil/bungee/SetupPlatformless.kt
/**
 * MV-Util
 * Copyright (C) 2020 Mariell Hoversholm, Nahuel Dolores
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package us.minevict.mvutil.bungee

import co.aikar.commands.BungeeCommandIssuer
import net.md_5.bungee.api.ProxyServer
import net.md_5.bungee.api.plugin.Plugin
import us.minevict.mvutil.bungee.ext.copyResource
import us.minevict.mvutil.common.utils.Platformless
import java.util.concurrent.CompletableFuture

/**
 * @since 5.0.0
 */
internal object SetupPlatformless {
    fun setup(plugin: MinevictusUtilsBungee) {
        Platformless.messageSender = { issuer, components ->
            if (issuer is BungeeCommandIssuer) {
                issuer.issuer.sendMessage(*components)
            }
        }
        Platformless.asyncRunner = {
            val future = CompletableFuture<Any?>()
            ProxyServer.getInstance().scheduler.runAsync(plugin) {
                try {
                    future.complete(it())
                } catch (ex: Throwable) {
                    future.completeExceptionally(ex)
                }
            }
            future
        }
        Platformless.exportJarResource = b@{ plug, name, file, generate ->
            if (plug is Plugin) plug.copyResource(name, file, failSafe = true, generate = generate)
            else false
        }
    }
}